File: backend/src/cars/api/dto/car.dto.ts
import { ApiProperty } from '@nestjs/swagger';
import { IsInt, IsNotEmpty, IsOptional, IsString, Min } from 'class-validator';
import { Type } from 'class-transformer';
import { MaxModelYear } from './isModelYear.decorator';
import { Car } from '../../domain/entities/car.entity';

export class CarDto {
  @IsString()
  @IsNotEmpty()
  @ApiProperty()
  id: string;

  @IsString()
  @IsNotEmpty()
  @ApiProperty()
  brand: string;

  @IsString()
  @IsNotEmpty()
  @ApiProperty()
  model: string;

  @IsInt()
  @Type(() => Number)
  @Min(1886)
  @MaxModelYear()
  @ApiProperty()
  year: number;

  @IsString()
  @IsNotEmpty()
  @ApiProperty()
  licensePlate: string;

  @IsOptional()
  @IsString()
  @ApiProperty({ type: String, nullable: true })
  color: string | null;

  @IsOptional()
  @IsInt()
  @Type(() => Number)
  @Min(0)
  @ApiProperty({ type: Number, nullable: true })
  mileage: number | null;

  constructor(car?: Car) {
    if (car) {
      this.id = car.id;
      this.brand = car.brand;
      this.model = car.model;
      this.year = car.year;
      this.licensePlate = car.licensePlate;
      this.color = car.color;
      this.mileage = car.mileage;
    }
  }

  static fromDomain(car: Car): CarDto {
    return new CarDto(car);
  }

  toDomain(): Car {
    return {
      id: this.id,
      brand: this.brand,
      model: this.model,
      year: this.year,
      licensePlate: this.licensePlate,
      color: this.color,
      mileage: this.mileage,
      deletedAt: null,
    };
  }
}

File: frontend/lib/features/vehicles/data/models/car_dto.dart
import 'package:autobook/features/vehicles/domain/entities/car.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part 'car_dto.freezed.dart';
part 'car_dto.g.dart';

@freezed
abstract class CarDto with _$CarDto {
  const CarDto._();

  const factory CarDto({
    required String id,
    required String brand,
    required String model,
    required int year,
    required String licensePlate,
    String? color,
    int? mileage,
  }) = _CarDto;

  factory CarDto.fromJson(Map<String, dynamic> json) => _$CarDtoFromJson(json);

  factory CarDto.fromDomain(Car car) => CarDto(
        id: car.id,
        brand: car.brand,
        model: car.model,
        year: car.year,
        licensePlate: car.licensePlate,
        color: car.color,
        mileage: car.mileage,
      );

  Car toDomain() => Car(
        id: id,
        brand: brand,
        model: model,
        year: year,
        licensePlate: licensePlate,
        color: color,
        mileage: mileage,
      );
}

@freezed
abstract class CarsListResponse with _$CarsListResponse {
  const factory CarsListResponse({required List<CarDto> cars}) =
      _CarsListResponse;

  factory CarsListResponse.fromJson(Map<String, dynamic> json) =>
      _$CarsListResponseFromJson(json);
}
